File: Ahorn/entities/FeatherBumper.jl
module AnonhelperFeatherBumper

using ..Ahorn, Maple
@mapdef Entity "Anonhelper/FeatherBumper" FeatherBumper(x::Integer, y::Integer, Wobble::Bool=true)
const placements = Ahorn.PlacementDict(
    "Feather Bumper (Anonhelper)" => Ahorn.EntityPlacement(
        FeatherBumper
    )
)

Ahorn.nodeLimits(entity::FeatherBumper) = 0, 1

sprite = "objects/AnonHelper/featherBumper/plugin.png"

function Ahorn.selection(entity::FeatherBumper)
    x, y = Ahorn.position(entity)
    nodes = get(entity.data, "nodes", ())

    if !isempty(nodes)
        nx, ny = Int.(nodes[1])

        return [Ahorn.getSpriteRectangle(sprite, x, y), Ahorn.getSpriteRectangle(sprite, nx, ny)]
    end

    return Ahorn.getSpriteRectangle(sprite, x, y)
end

function Ahorn.renderSelectedAbs(ctx::Ahorn.Cairo.CairoContext, entity::FeatherBumper)
    x, y = Ahorn.position(entity)
    nodes = get(entity.data, "nodes", ())

    if !isempty(nodes)
        nx, ny = Int.(nodes[1])

        theta = atan(y - ny, x - nx)
        Ahorn.drawArrow(ctx, x, y, nx + cos(theta) * 8, ny + sin(theta) * 8, Ahorn.colors.selection_selected_fc, headLength=6)
        Ahorn.drawSprite(ctx, sprite, nx, ny)
    end
end

Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::FeatherBumper, room::Maple.Room) = Ahorn.drawSprite(ctx, sprite, 0, 0)

end

File: Loenn/entities/feather_bumper.lua
local utils = require("utils")

local featherBumper = {}

featherBumper.name = "Anonhelper/FeatherBumper"
featherBumper.depth = 0
featherBumper.nodeLineRenderType = "line"
featherBumper.texture = "objects/AnonHelper/featherBumper/Idle22"
featherBumper.nodeLimits = {0, 1}
featherBumper.placements = {
    name = "feather_bumper",
    data = {
        wobble = true,
    }
}

function featherBumper.selection(room, entity)
	local ex, ey = entity.x or 0, entity.y or 0
    local main = utils.rectangle(ex - 11 , ey - 11, 22, 22)

    if entity.nodes then
        local nodeSelections = {}
        for _, node in ipairs(entity.nodes) do
			local nx, ny = entity.x or 0, entity.y or 0
            table.insert(nodeSelections, utils.rectangle(nx - 11, ny -11, 22, 22))
        end
        return main, nodeSelections
    end

    return main, { }
end

return featherBumper